### DISCUSSION TITLE
How to access context via MarkpExtension.ProvideValue

### DISCUSSION THREAD

**Discussion by mwpowellhtx** [dotnet/wpf#9000] [Q&A]
Apparently with [`MarkupExtension.ProvideValue`](https://github.com/dotnet/wpf/blob/main/src/Microsoft.DotNet.Wpf/src/System.Xaml/System/Windows/Markup/MarkupExtension.cs#L26) being the key, in particular `IServiceProvider serviceProvider`...

Need some basic guidance on how to gain insights into a particular usage context, the control, the outermost XAML element, and any _resources_ that might be had in this context, i.e. `Control.Resources`, `Windows.Resources`, if at all possible.

For background, I am [_scanning through the Learn docs_](https://learn.microsoft.com/en-us/dotnet/desktop/xaml-services/service-contexts-with-type-converters-and-markup-extensions), but nothing is really standing out as an obvious _this is the way_ moment, to be honest. More probably, I have just not asked the right question yet, but I am getting there.

**Comment by mwpowellhtx** [comment]
To clarify, in the context of a `MarkupExtension` usage, my goal is to identify, if possible, the nearest possible `<*.Resources/>`, either the _control resources_, or outermost `<Window.Resources/>`, etc, possible, of a particular type. Surely something like this must be possible given the correct service provider, navigating the XAML element plumbing at the CSharp object level.

**Comment by mwpowellhtx** [comment]
I wonder, for instance, is `IAmbientProvider` something of interest, considering [`ResourceDictionary`](https://learn.microsoft.com/en-us/dotnet/api/system.windows.frameworkelement.resources) supporting `FrameworkElement.Resources` is annotated:

```csharp
[System.Windows.Markup.Ambient]
public System.Windows.ResourceDictionary Resources { get; set; }
```

Much of that API is geared toward knowing the key, usually a string, via _indexing operator_, or `FindResource` method. However, I do not necesssarily want to know the string, rather, identify the value by its type.

**Comment by miloush** [comment]
Sorry I missed this question. Did you see how [`StaticResourceExtension`](https://github.com/dotnet/wpf/blob/504fd82cd5acb1844141f58fc65eb1445e810ce7/src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/StaticResourceExtension.cs#L202) does it?
